### PR TITLE
E2E Automation framework changes

### LINKED ISSUE #238: E2E framework for kubeedge
**Why is this needed**:
E2E framework should be developed for testing of resurce deployments in kubeedge, and also performance tests with Cloud and Edge nodes

### PR DESCRIPTION
**Pull request by pavan187**
**What type of PR is this?**
 /kind test

**What this PR does / why we need it**:
PR has the changes of E2E automation framework development.

**Which issue(s) this PR fixes**: 
E2E framework added to test the end-to-end testing of application deployment using kube client, verify the application deployments are happening successfully, This framework can also be utilized for  performance tests.

Fixes #238

### CODE REVIEW

File: e2e/integration/scripts/execute.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    echo "edge_core process is killed"
+fi
+
+PWD=${curpath}/e2e/integration
+sudo rm -rf $PWD/deployment/deployment.test
+go get github.com/onsi/ginkgo/ginkgo
+sudo cp $GOPATH/bin/ginkgo /usr/bin/
+# Specify the module name to compile in below command
+bash -x $PWD/scripts/compile.sh $1
+export MASTER_IP=121.244.95.60
```
Comment (sids-b) [inline e2e/integration/scripts/execute.sh]: We need to find a mechanism to mask the IP address of the master. If our Kubernetes master is public, anyone can use it and our tests may fail due to load. Can we try to find out a way to masks the IP ??

File: e2e/integration/scripts/cleanup.sh [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
```
Comment (sids-b) [inline e2e/integration/scripts/cleanup.sh]: LICENSE head please

File: e2e/integration/scripts/execute.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Comment (sids-b) [inline e2e/integration/scripts/execute.sh]: LICENSE head please

File: e2e/integration/deployment/deployment_test.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/apps/v1"
+	metav1 "k8s.io/api/core/v1"
+)
+
+const (
+	AppHandler    = "/api/v1/namespaces/default/pods"
+	NodeHandler   = "/api/v1/nodes"
+	Deployment    = "/apis/apps/v1/namespaces/default/deployments"
```
Comment (sids-b) [inline e2e/integration/deployment/deployment_test.go]: Deployment -> DeploymentHandler ? to keep the naming consistent

File: e2e/integration/helpers/helpers.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+		utils.Failf("HTTP Response reading has failed: %v", err)
+		return nodestatus
+	}
+	err = json.Unmarshal(contents, &node)
+	if err != nil {
+		utils.Failf("Unmarshal HTTP Response has failed: %v", err)
+		return nodestatus
+	}
+
+	fmt.Println(string(node.Status.Phase))
```
Comment (sids-b) [inline e2e/integration/helpers/helpers.go]: print required ?

File: e2e/integration/scripts/setup.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        exit 1
+    fi
+    sed -i "s|master: .*|master: ${MASTER_IP}|g" ${CLOUD_PATH}
+    sed -i "s|ca: .*|ca: tmp/rootCA.crt|g" ${CLOUD_PATH}
+    sed -i "s|cert: .*|cert: tmp/kubeedge.crt|g" ${CLOUD_PATH}
+    sed -i "s|key: .*|key: tmp/kubeedge.key|g" ${CLOUD_PATH}
+}
+
+create_edge_config
+create_cloud_config
```
Comment (m1093782566) [inline e2e/integration/scripts/setup.sh]: a newline please.

File: e2e/integration/scripts/generate_cert.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+echo "---------------------------------------------"
+
+#Generate temparory folder in Edge and Cloud folders to copy certs
+sudo mkdir -p $GOPATH/src/github.com/kubeedge/kubeedge/edge/tmp
+sudo mkdir -p $GOPATH/src/github.com/kubeedge/kubeedge/cloud/edgecontroller/tmp
+#Copy the generated certs to respective paths
+sudo cp -r rootCA.crt rootCA.key kubeedge.crt kubeedge.key $GOPATH/src/github.com/kubeedge/kubeedge/edge/tmp
+sudo cp -r rootCA.crt rootCA.key kubeedge.crt kubeedge.key $GOPATH/src/github.com/kubeedge/kubeedge/cloud/edgecontroller/tmp
+
+echo "-----Certificate are Copied to Edge and Cloud Nodes-----"
```
Comment (m1093782566) [inline e2e/integration/scripts/generate_cert.sh]: a newline please.

File: e2e/integration/helpers/helpers.go [hunk 8]
```diff
@@ -0,0 +1,341 @@
+/*
```
Comment (m1093782566) [inline e2e/integration/helpers/helpers.go]: helpers/helpers.go -> helpers.go? For the sake of reducing folder depth.

File: e2e/integration/helpers/helpers.go [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+	output, err := cmd.CombinedOutput()
+	if err != nil{
+		utils.Failf("CombinedOutput failed", err)
+		return err
+	}
+	if len(output) > 0{
+		fmt.Printf("=====> Output: %s\n", string(output))
+	}
+	return nil
+}
```
Comment (m1093782566) [inline e2e/integration/helpers/helpers.go]: A newline please.

**Comment by m1093782566** [comment]
cc @sujithsimon22 Please help review.

File: e2e/integration/helpers/helpers.go [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+)
+
+const (
+	podLabelSelector = "?labelSelector="
+)
+
+func SenHttpRequest(method, reqapi string) (error, *http.Response){
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description explaining the functionality of this function

Comment (sujithsimon22) [reply e2e/integration/helpers/helpers.go]: nit: SenHttpRequest -> SendHttpRequest ?

File: e2e/integration/helpers/helpers.go [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+	if err != nil {
+		// handle error
+		utils.Failf("HTTP request is failed :%v", err)
+		return err, resp
+	}
+
+	return nil, resp
+}
+
+func GetDeployments(list *metav1.DeploymentList, getDeploymentApi string) error{
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+	err = json.Unmarshal(contents, &list)
+	if err != nil {
+		utils.Failf("Unmarshal HTTP Response has failed: %v", err)
+		return err
+	}
+	return nil
+
+}
+
+func GenerateDeploymentBody(deploymentName, image string, port, replica int) (error, map[string]interface{}){
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+	resp, err := client.Do(req)
+	utils.InfoV6("%s %s %v in %v", req.Method, req.URL, resp.Status, time.Now().Sub(t))
+	if err != nil {
+		// handle error
+		utils.Failf("HTTP request is failed :%v", err)
+		return false
+	}
+	return true
+}
+func DeleteDeployment(DeploymentApi, deploymentname string)int{
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 14]
```diff
@@ -0,0 +1,341 @@
...
+	}
+	err = json.Unmarshal(contents, &pods)
+	if err != nil {
+		utils.Failf("Unmarshal HTTP Response has failed: %v", err)
+		return pods, nil
+	}
+	return pods, nil
+}
+
+func GetPodState(apiserver string) (string, int){
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+			if statusCode == 404{
+				count ++
+			}
+		}
+		return count
+	}, "120s", "4s").Should(Equal(len(podlist.Items)), "Delete Application deployment is Unsuccessfull, Pod has not come to Running State")
+
+}
+
+func GenerateNodeReqBody(nodeid string) (error, map[string]interface{}){
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+		return err
+	}
+	utils.InfoV6("%s %s %v in %v", req.Method, req.URL, resp.Status, time.Now().Sub(t))
+	defer resp.Body.Close()
+
+	Expect(resp.StatusCode).Should(Equal(http.StatusCreated))
+	return nil
+}
+
+func CheckNodeReadyStatus(ctx *utils.TestContext, nodehandler, nodename string) string{
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+	if err != nil {
+		utils.Failf("Unmarshal HTTP Response has failed: %v", err)
+		return nodestatus
+	}
+
+	fmt.Println(string(node.Status.Phase))
+	return string(node.Status.Phase)
+}
+
+func CheckNodeDeleteStatus(ctx *utils.TestContext, nodehandler, nodename string) int{
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 18]
```diff
@@ -0,0 +1,341 @@
...
+	err, resp := SenHttpRequest(http.MethodGet, ctx.Cfg.ApiServer+nodehandler+ "/" + nodename)
+	if err != nil{
+		utils.Failf("Sending SenHttpRequest failed: %v", err)
+		return -1
+	}
+	defer resp.Body.Close()
+	return  resp.StatusCode
+}
+
+func PrintCombinedOutput(cmd *exec.Cmd) error{
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/utils/config.go [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+//Get config path from Env or hard code the file path
+func getConfigPath() string {
+	path := os.Getenv("TESTCONFIG")
+	if path == "" {
+		path = "config.json"
+	}
+	return path
+}
+
+func RandomInt(min, max int) int {
```
Comment (sujithsimon22) [inline e2e/integration/utils/config.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/utils/context.go [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+//function to get the testcontext Object.
+func NewTestContext(cfg Config) *TestContext {
+	return &TestContext{
+		Cfg: cfg,
+	}
+}
+
+func MapLables(ls map[string]string) string {
```
Comment (sujithsimon22) [inline e2e/integration/utils/context.go]: Pls add a comment with a short description above this function explaining its functionality

File: e2e/integration/helpers/helpers.go [hunk 21]
```diff
@@ -0,0 +1,341 @@
...
+const (
+	podLabelSelector = "?labelSelector="
+)
+
+func SenHttpRequest(method, reqapi string) (error, *http.Response){
+	var body io.Reader
+	var resp *http.Response
+
+	client := &http.Client{}
+	req, err := http.NewRequest(method, reqapi, body)
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Can we use camel case for "reqapi" ?

File: e2e/integration/helpers/helpers.go [hunk 22]
```diff
@@ -0,0 +1,341 @@
...
+	var err error
+	var body io.Reader
+
+	client := &http.Client{}
+	switch operation {
+	case "POST":
+		Port := utils.RandomInt(10, 100)
+		err, body := GenerateDeploymentBody(UID, ImageUrl, Port, replica)
+		if err != nil {
+			utils.Failf("Payload marshalling failed: %v", err)
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: "Payload marshalling failed"  -> Is this correct ? Aren't we performing Unmarshal operation  inside GenerateDeploymentBody() ??

File: e2e/integration/helpers/helpers.go [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+func DeleteDeployment(DeploymentApi, deploymentname string)int{
+	err, resp := SenHttpRequest(http.MethodDelete, DeploymentApi+"/"+deploymentname)
+	if err != nil {
+		// handle error
+		utils.Failf("HTTP request is failed :%v", err)
+		return -1
+	}
+
+	defer resp.Body.Close()
+
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Please remove newline

File: e2e/integration/helpers/helpers.go [hunk 24]
```diff
@@ -0,0 +1,341 @@
...
+	}
+	return pods, nil
+}
+
+func GetPodState(apiserver string) (string, int){
+	var pod v1.Pod
+
+	err, resp := SenHttpRequest(http.MethodGet, apiserver)
+	if err != nil {
+		utils.Failf("GetPodState :SenHttpRequest failed: %v", err)
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: nit : SenHttpRequest -> SendHttpRequest

File: e2e/integration/helpers/helpers.go (hunk 16 above)
Comment (sujithsimon22) [reply e2e/integration/helpers/helpers.go]: Can we pls use camel case for nodehandler, nodename ??

File: e2e/integration/helpers/helpers.go [hunk 25]
```diff
@@ -0,0 +1,341 @@
...
+	utils.InfoV6("%s %s %v in %v", req.Method, req.URL, resp.Status, time.Now().Sub(t))
+	defer resp.Body.Close()
+
+	Expect(resp.StatusCode).Should(Equal(http.StatusCreated))
+	return nil
+}
+
+func CheckNodeReadyStatus(ctx *utils.TestContext, nodehandler, nodename string) string{
+	var node v1.Node
+	var nodestatus ="unknown"
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: Can we pls use camel case for "nodestatus" ?

File: e2e/integration/helpers/helpers.go [hunk 26]
```diff
@@ -0,0 +1,341 @@
...
+	err, resp := SenHttpRequest(http.MethodGet, ctx.Cfg.ApiServer+nodehandler+ "/" + nodename)
+	if err != nil{
+		utils.Failf("Sending SenHttpRequest failed: %v", err)
```
Comment (sujithsimon22) [inline e2e/integration/helpers/helpers.go]: nit: SenHttpRequest -> SendHttpRequest

File: e2e/integration/utils/context.go (hunk 20 above)
Comment (sujithsimon22) [reply e2e/integration/utils/context.go]: nit: MapLables -> MapLabels ??

File: e2e/integration/helpers/helpers.go [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	defer resp.Body.Close()
+
+	return resp.StatusCode
+}
+
+//fucntion to show the os command injuction in combined format
+func PrintCombinedOutput(cmd *exec.Cmd) error{
+	fmt.Printf("===========> Executing: %s\n", strings.Join(cmd.Args, " "))
```
Comment (kevin-wangzefeng) [inline e2e/integration/helpers/helpers.go]: Should not use fmt.Printf, need a unified logging or output method

File: e2e/integration/utils/node.go [hunk 28]
```diff
@@ -0,0 +1,124 @@
...
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+)
+
+//Function to get the pods from Edged
+func DeRegisterNodeFromMaster(ctx *TestContext, nodehandler, nodename string) error {
```
Comment (edisonxiang) [inline e2e/integration/utils/node.go]: please update the description for this function

File: e2e/integration/utils/node.go [hunk 29]
```diff
@@ -0,0 +1,124 @@
...
+	if err!= nil{
+		Failf("Unmarshal body failed: %v", err)
+		return err, temp
+	}
+
+	return nil, temp
+}
+
+//Function to get the pods from Edged
+func RegisterNodeToMaster(ctx *TestContext, UID, nodehandler string) error {
```
Comment (edisonxiang) [inline e2e/integration/utils/node.go]: ditto

File: Makefile [hunk 30]
```diff
@@ -29,3 +29,7 @@ edge_small_build:
 edgecontroller:
 	cd cloud/edgecontroller && $(MAKE)
 
+.PHONY: e2e_test
+e2e_test:
+	bash e2e/integration/scripts/execute.sh
```
Comment (m1093782566) [inline Makefile]: please remove the integration here for the sake of avoiding confusing.

File: e2e/integration/deployment/deployment_suite_test.go [hunk 31]
```diff
@@ -0,0 +1,100 @@
+/*
```
Comment (m1093782566) [inline e2e/integration/deployment/deployment_suite_test.go]: deployment_suite_test.go -> deployment_test.go

File: e2e/integration/deployment/deployment_test.go [hunk 32]
```diff
@@ -0,0 +1,81 @@
+/*
```
Comment (m1093782566) [inline e2e/integration/deployment/deployment_test.go]: Can we merge this file into the above file?

File: e2e/integration/scripts/fast_test [hunk 33]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash -x
```
Comment (m1093782566) [inline e2e/integration/scripts/fast_test]: fast_test -> fast_test.sh?

File: Makefile (hunk 30 above)
Comment (m1093782566) [reply Makefile]: And any e2e test related files should be placed in tests/e2e/...
